src/python/pants/backend/codegen/protobuf/python/rules.py

```diff
@@ -96,8 +98,25 @@ async def generate_python_from_protobuf(
             output_directories=(output_dir,),
         ),
     )
-    normalized_snapshot = await Get(Snapshot, RemovePrefix(result.output_digest, output_dir))
-    return GeneratedSources(normalized_snapshot)
+
+    # We must do some path manipulation on the output digest for it to look like normal sources,
+    # including adding back the original source root.
```

Comment:
It's likely fine either way (and easy to change later), but to avoid colliding with files that exist on disk, it might be safer to locate these in a synthetic source root of some sort. Historically in v1 this happened because when we generated code we did so under `.pants.d`, meaning the generated code always had a long prefix that would collide with real code.

Reply:
Hm, the challenge there is that we would need to teach the source roots rule what the synthetic root is. Maybe we would use a global constant?

For now, I'm not expecting any collisions because Protobuf adds a `_pb2.py` to the end of every file, so a user would really need to go out of there way to already have a file in the same location.